File: Ruby/Euler008.rb
#!/usr/bin/env ruby

def main
  factor_ct = 13
  ord0 = '0'.ord

  contents = File.open('input/Euler008.txt') do |f|
    f.read.split('').filter { |c| '0' <= c && c <= '9' }.map { |c| c.ord - ord0 }
  end

  buffer = factor_ct.times.map { 0 }
  product = 0

  contents.each_with_index { |val, idx|
    buffer[idx % factor_ct] = val
    product = [product, buffer.reduce(1, :*)].max
  }

  puts product
end

if __FILE__ == $0
  main
end

File: Python/Euler008.py
#!/usr/bin/env python3
from __future__ import print_function
from functools import reduce
from math import sqrt
from operator import mul

def main():
    factor_ct = 13
    char0 = 48

    with open('input/Euler008.txt') as f:
        contents = [ord(c) - char0 for c in f.read() if '0' <= c <= '9']

    buffer = [0] * factor_ct

    product = 0
    for index, value in enumerate(contents):
        buffer[index % factor_ct] = value
        product = max(product, reduce(mul, buffer))

    print(product)


if __name__ == '__main__':
    main()
